Fix: libpe_status, xml: Support integer as rule type attribute

nrwahl2 opened the pull request:
The documentation specifies three potential values for the type 
attribute of rule expressions:

- string
- integer
- version

Yet the rule.rng schema and pe__eval_attr_expr() expect "number" instead 
of "integer".

This pull corrects the documentation to align with the schema and code.

Thanks @tomjelinek for reporting this.

---

Edited summary:

This pull does the following:
- Creates a pcmk_parse_double() function (strtod() wrapper) and unit tests
- Creates a pcmk__char_in_any_str() function (multi-strchr() wrapper) and unit tests
- Adds "integer" to the rule schema as a valid type attribute
- Adds "integer" support to libpe_status, updates "number" to specify a floating-point comparison, and uses "integer" for integer comparisons. (Previously, "number" specified an integer comparison.)
- Updates the documentation to align with the schema and libpe_status

nrwahl2 commented:
Alternatively, would it make sense to parse the values with strtod() and treat them as doubles?

kgaillot commented:
I'm thinking we should leave the documentation as-is, and modify the schema and code to accept either "integer" or "number" interchangeably. Perhaps at a future Pacemaker major-version bump we can make "number" behave as a floating point.

To modify the schema, see xml/Readme.md. This will be a compatible change.

kgaillot commented:
Another possibility would be to treat "number" as a double if it contains a '.', and an integer otherwise. That would keep backward compatibility for anyone using it as an integer currently while allowing users to start using it for floating points. In that case the documentation, code and schema could be brought in line as:
- "integer", or "number" with a value without '.': integer
- "number" with a value with '.': double-precision floating point

nrwahl2 commented:
> To modify the schema, see xml/Readme.md. This will be a compatible change.

Ack. As we discussed in chat, the source of confusion was the use of an old rule schema in the current constraints schema.

---

> Another possibility would be to treat "number" as a double if it contains a '.', and an integer otherwise. That would keep backward compatibility for anyone using it as an integer currently while allowing users to start using it for floating points.

Would this actually do anything to alleviate your concern in the RHBZ? You said:

> If anyone's using it, they're using "number" with integers, and they may reasonably expect values not to have any of the unusual corner cases of floating-point arithmetic, so we should keep "number" syntax and integer behavior.

If they're currently using integer strings (i.e., without a '.'), then they shouldn't be prone to any floating-point corner cases AFAIK. The floating point promotion would just add zeroes after the decimal place anyway.

Users would only hit a floating-point corner case (e.g., a comparison failing due to a precision issue in FP arithmetic) after the suggested change if they're **currently** using floating-point strings (i.e., with a '.') and relying on Pacemaker to truncate the value to an int. Then after the change, those values would be parsed as doubles and may compare incorrectly, in theory.

So it seems to me that parsing numbers with '.' as floating-point would be just the same as parsing ALL numbers as floating-point, when it comes to the risk of hitting unexpected behavior in a corner case.

nrwahl2 commented on cts/cli/regression.tools.exp:
```diff
@@ -1108,8 +1108,8 @@ Error performing operation: Node not found
 
- dummy	(ocf::pacemaker:Dummy):	Stopped
- Fence	(stonith:fence_true):	Stopped
+ dummy	(ocf::pacemaker:Dummy):	 Stopped
```
These extra spaces are from something @clumens added IIRC. They shouldn't make a difference since CTS's diff ignores whitespace. I figured I'd go ahead and add this .exp file with the others. Let me know if you'd rather I remove it.

kgaillot replied:
This is fine

kgaillot commented:
> So it seems to me that parsing numbers with '.' as floating-point would be just the same as parsing ALL numbers as floating-point, when it comes to the risk of hitting unexpected behavior in a corner case.

A bit contrived, but:

    #include <stdio.h>
    #include <stdlib.h>


    int main(int argc, char **argv)
    {
        unsigned long integer = 9223372036854775809UL;
        double number = strtod("9223372036854775809.0", NULL);

        printf("%lu %f\n", integer, number);
        return 0;
    }

However that wouldn't affect our situation since we currently parse the values as 32-bit integers (which I didn't realize until checking just now), so technically you're right. :) Which makes me think we should indeed parse "number" as double, and start parsing "integer" with crm_parse_ll(). That would clean things up nicely and shouldn't affect any existing usage adversely.

nrwahl2 commented:
> A bit contrived, but:

The best corner cases are contrived :)
And ah yes, large ints that require high precision. I forgot about losses when converting those to doubles.

> That would clean things up nicely and shouldn't affect any existing usage adversely.

Ack. So I take it we're not worrying about any odd use cases where the expected result of a comparison depends on truncation to int. I support going with double for "number" as long as you're good with it, since it is more flexible.

I was working on a pcmk__scan_double() and pcmk_parse_double(). I back-burnered them to focus on the meat of this issue, after having some initial difficulty detecting overflow and underflow in a way that's reliable on both ANSI and C99. (The strtod() return value for underflow is 0 and ERANGE on ANSI, while it's "<= smallest positive normalized number and **maybe** ERANGE (implementation-dependent)" on C99.)

If we've settled on parsing "number" as double, I'll get back to work on that and add the result to the PR.

kgaillot commented:
> Ack. So I take it we're not worrying about any odd use cases where the expected result of a comparison depends on truncation to int. I support going with double for "number" as long as you're good with it, since it is more flexible.

Yep. Using a "." and expecting an integer comparison would be twisted enough that I'd consider it a bug.

> I was working on a pcmk__scan_double() and pcmk_parse_double(). I back-burnered them to focus on commits for the main focus of this issue, after having some initial difficulty detecting overflow and underflow in a way that's reliable on both ANSI and C99. (The strtod() return value for underflow is 0 and ERANGE on ANSI, while it's "<= smallest positive normalized number and **maybe** ERANGE (implementation-dependent)" on C99.)
> 
> If we've settled on parsing "number" as double, I'll get back to work on that and add the result to the PR.

Sounds good

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,129 @@ crm_parse_int(const char *text, const char *default_text)
+ * \param[out] end_text  If not NULL, where to store pointer to just
+ *                       after value
+ *
+ * \return Standard Pacemaker return code (also set errno on error)
```
We could make the doc comment more helpful by specifying the values of errno that might be set -- namely, EINVAL for bad input and ERANGE for overflow/underflow. We can do the same for scan_ll(), crm_parse_ll(), crm_parse_int(), and pcmk_parse_double().

Would those details go in the \return section?

nrwahl2 replied:
Come to think of it, returning anything at all seems redundant, since rc is always equal to errno. It seems that one or the other would be fine. I used both because scan_ll() uses both.

kgaillot replied:
scan_ll() sets errno for backward API compatibility, but going forward for new functions we only need to use the standard return code.

Specific values can be documented in either \return or a \note. See pcmk__remote_ready() for a trivial example. A series of \retval entries is also fine if all values are enumerated, but since we're still transitioning between the legacy and standard rc's I prefer \return so we can say which it is.

nrwahl2 commented:
This has grown pretty large. I think I'm finished, until there's feedback on things that need to be changed.

A lot of the commits are prep work (e.g., adding new utilities and their tests, copying schemas, etc.). Four of them are minor issues and opportunities I happened upon while doing the prep work. The last few commits directly address the issue.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -847,6 +847,44 @@ pcmk__str_none_of(const char *s, ...)
+     * Passing a char to va_start() can generate compiler warnings,
+     * so ch is declared as an int.
+     */
+    va_start(ap, ch);
```
I think this is fine, since strchr()'s prototype accepts an int already and converts it to a char. I had to make this change after Clang refused to compile.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,129 @@ crm_parse_int(const char *text, const char *default_text)
+ * \return Standard Pacemaker return code (also set errno on error)
+ */
+static int
+pcmk__scan_double(const char *text, double *result, char **end_text)
```
Should static functions use the "pcmk__" prefix or not (since there's no possibility of name conflicts outside the file)? I noticed that the scan_ll() name was set only 6 months ago as a best-practices commit: fc94233e.

kgaillot replied:
It's not terribly important either way, but I prefer not. It's not so obvious at the moment since we have so many legacy names, but eventually it would be: pcmk_ = public, pcmk__ = private, none = static, which would be a quick visual clue when you're browsing code. Also, as you mentioned, the prefixes are needed primarily to avoid name collisions, which is not an issue with static functions and just makes their names unnecessarily longer.

nrwahl2 replied:
Ack, next push will use scan_double().

nrwahl2 replied:
Next push will also update the macros in the pcmk_parse_double() unit test to remove "PCMK__", since they're local and not in a header file.

nrwahl2 commented:
One other thing: Does any of this require a feature set bump?

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -17,7 +17,9 @@
 #include <ctype.h>
+#include <float.h>  // DBL_MIN
 #include <limits.h>
+#include <math.h>   // fabs()
```
Thanks for the heads-up there. I'll try to keep that in mind should it become relevant :)

> the new name is CRM_SCORE_INFINITY

Surprised we're not using a PCMK prefix for that too.

kgaillot replied:
The approach has evolved over the years. CRM_SCORE_INFINITY was part of an early attempt to use existing prefixes more consistently. Since then we've decided on PCMK (eventually for everything, but starting with new or internal stuff in libcrmcommon and libpacemaker).

My goal is to rename all internal stuff first, then eventually rename public API with the legacy names as compatibility wrappers. Coccinelle should make that easier. For example this case would become

    #define PCMK_SCORE_INFINITY 1000000
    ...
    // Begin compatibility section not accessible by pacemaker code
    ....
    //! \deprecated Use PCMK_SCORE_INFINITY instead
    #define CRM_SCORE_INFINITY PCMK_SCORE_INFINITY

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,129 @@ crm_parse_int(const char *text, const char *default_text)
+
+        } else if (errno != 0) {
+            rc = errno;
+            local_result = -1;
```
I don't think so. strtod() sets local_result = 0 if it can't parse the string. This line sets it back to -1.

And the initialization of local_result to -1.0 remains useful in the event that (text == NULL) and we don't enter this main block. We would have to either initialize to -1.0 as we do now, or set it to -1.0 in an else block (i.e., text == NULL).

kgaillot replied:
Whoops, my bad, I missed that.

nrwahl2 replied:
No problem. It's easy to miss, so good thing you brought it up for me to comment the reason for re-setting it.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,129 @@ crm_parse_int(const char *text, const char *default_text)
+
+            for (const char *p = text; p != local_end_text; p++) {
+                if (strchr("0.eE", *p) == NULL) {
+                    // TODO: Handle hex doubles
```
Haha, ack, it'll be removed on the next push. I thought we might eventually want that generalizability. Kind of hard to imagine using it though.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,129 @@ crm_parse_int(const char *text, const char *default_text)
+             * also sets local_result == 0, so a parse failure would
+             * match this test condition prematurely.
+             */
+            rc = errno = ERANGE;
```
Yes, thank you.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,129 @@ crm_parse_int(const char *text, const char *default_text)
+            crm_warn("Floating-point value parsed from '%s' would %sflow "
+                     "(using %g instead)", text,
+                     (fabs(local_result) > DBL_MIN) ? "over" : "under",
+                     local_result);
```
> I think this (and later messages) should be at debug level. It's better for library functions to stick to info level or lower, if they can return a sufficiently informational return code, for two reasons

Ack, I agree with that. I'll crm_debug() the "errors" and crm_trace() the success.

> Exceptions would be legacy functions that don't return a good enough (or any) return code

I modeled these err/warn messages after scan_ll(). Say what you will about scan_ll(), but its return codes are sufficient. Do you think we should drop to crm_debug() in scan_ll()?

---

I've also removed the "ANSI" note from the comment (next push), since we're not really making an effort at compatibility with the 30-year-old standard. The underflow note will have the C99 paragraph without "C99: ".

kgaillot replied:
> I modeled these err/warn messages after scan_ll(). Say what you will about scan_ll(), but its return codes are sufficient. Do you think we should drop to crm_debug() in scan_ll()?

scan_ll() has to keep the logs (and set errno) because it's used in (public API) crm_parse_ll(), which only returns -1 on error. We're often stuck with such "exceptions", at least until the next public API compatibility break (my rough goal is to keep at least 6 years between them, to minimize the disruption for distros).

nrwahl2 commented on lib/common/tests/strings/pcmk_parse_double.c:
```diff
@@ -0,0 +1,135 @@
+    // With default_text
+    PCMK__ASSERT_DBL_EQ(pcmk_parse_double(NULL, "2.0"), 2.0);
+
+    PCMK__ASSERT_DBL_EQ(pcmk_parse_double("", "2.0"), PCMK__PARSE_DBL_DEFAULT);
```
I meant to ask you about that, actually. The above mirrors the behavior of crm_parse_ll() -- it uses the default text if text is NULL, but returns -1 if text is an empty string.

It makes sense from the perspective of how it handles string parsing in general. If you give scan_ll() (or scan_double()) a string that starts with non-numeric characters so that it fails to parse the string, it returns -1. In that sense, the empty string is just a bad string that it can't parse.

But it might be more intuitive to handle an empty string in the same way that we handle a NULL string. That's what we do in most places and what we do with pcmk__str_empty().

With that said, I don't know if we can change scan_ll()/crm_parse_ll()'s empty-string behavior since crm_parse_ll() is public. And for consistency, I would expect pcmk_parse_double() to handle an empty string in the same way as crm_parse_ll().

kgaillot replied:
Yep, that's another situation backward compatibility often puts us in. I think the main factor should be: how would an empty string arrive at this function, and if that involves the user at any point, how would they expect it to behave? If an empty string argument would never involve the user, then it doesn't matter, and we can just do something reasonable on the caller end. Also, since it's private API, it's not terribly important to decide now. But when we get around to making the legacy function names wrappers for new functions, we could change the behavior in the new function and hack it back to the old way in the wrapper.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -75,8 +75,7 @@ scan_ll(const char *text, long long *result, char **end_text)
 
-        if ((end_text == NULL) && (local_end_text != NULL)
-            && (local_end_text[0] != '\0')) {
+        if ((end_text == NULL) && !pcmk__str_empty(local_end_text)) {
```
Hmm, it's probably uncommon but that might be good to do.

nrwahl2 commented on include/crm/common/util.h:
```diff
@@ -51,6 +51,7 @@ gboolean crm_is_true(const char *s);
 int crm_str_to_boolean(const char *s, int *ret);
 long long crm_parse_ll(const char *text, const char *default_text);
 int crm_parse_int(const char *text, const char *default_text);
+double pcmk_parse_double(const char *text, const char *default_text);
```
I could see either:
- keeping scan_double() as a helper function and converting pcmk_parse_double to an internal function that wraps scan_double(), or
- modifying scan_double()'s interface to make it a friendly internal function.

I wouldn't want to use scan_double() as an internal function with its current interface.

The reason is that scan_double() returns an error/success code and stores the parsed value in a pointer-to-double parameter. That's nice if we want to do something like setting the pointer to NULL if the parse fails (which we don't currently do). But it doesn't allow us to use the return value as an argument to another function, as we sometimes do with crm_parse_ll() and crm_parse_int(). To me, that's a handy capability. Otherwise we have to do the parse first, storing the result in some variable, and pass that variable to another function.

You also said in a different comment:
> going forward for new functions we only need to use the standard return code.

For the reasons given above (ease of passing the return value as a function argument), I think we should either:
- continue setting errno, or
- use a status argument similar to what waitpid() uses, to store the standard Pacemaker return code.

I lean toward the latter. I'll probably do something like that in the next push. Let me know what you think :)

kgaillot replied:
> Otherwise we have to do the parse first, storing the result in some variable, and pass that variable to another function.

Actually I prefer that -- implicitly handling errors inside a function and having the caller ignore it is a recurring source of bugs. I've seen more than one where it didn't matter for a function's initial use case but a later use really needed to be checking for errors, but since the easy interface was available, it was overlooked.

I prefer returning the code, rather than putting the code in an argument or errno, precisely because it does force the developer to think about error checking (as well as for consistency).

The only time I don't mind returning a result directly (instead of a standard code) is when the function can't fail (due to simplicity, or asserting on errors), or the interface is dictated by its usage with an external library (e.g. a glib callback).

nrwahl2 replied:
Ah, understandable concerns. Will be resolved in the next push.

nrwahl2 commented on lib/common/tests/strings/pcmk__char_in_any_str.c:
```diff
@@ -0,0 +1,44 @@
+empty_list(void)
+{
+    g_assert_false(pcmk__char_in_any_str('x', NULL));
+    g_assert_false(pcmk__char_in_any_str('\0', NULL));
```
Understandable concern but I think `g_assert_(true|false)` is the way to go here.

1. g_assert() and g_assert_true() run the same check (G_LIKELY(expr)) if G_DISABLE_ASSERT is not set. The only differences are the behavior if G_DISABLE_ASSERT is set, and a slight log message difference.

```
#define g_assert_true(expr)             G_STMT_START { \
                                             if G_LIKELY (expr) ; else \
                                               g_assertion_message (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
                                                                    "'" #expr "' should be TRUE"); \
                                        } G_STMT_END
#define g_assert_false(expr)            G_STMT_START { \
                                             if G_LIKELY (!(expr)) ; else \
                                               g_assertion_message (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
                                                                    "'" #expr "' should be FALSE"); \
                                        } G_STMT_END
...
#else /* !G_DISABLE_ASSERT */
...
#define g_assert(expr)                  G_STMT_START { \
                                             if G_LIKELY (expr) ; else \
                                               g_assertion_message_expr (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
                                                                         #expr); \
                                        } G_STMT_END
```

2. The big-endian issue that we hit with glib option parsing was due to some evil-ness in the way it assigns to arguments. It takes a pointer to the store variable and casts it to a pointer to gboolean when assigning. Since a gboolean is 4 bytes and a bool is 1 byte, we lose the least significant 3 bytes of the rvalue during assignment, when arg_data is a bool variable.
```
          *(gboolean *)entry->arg_data = !(entry->flags & G_OPTION_FLAG_REVERSE);
```

A normal cast from gboolean to bool (or vice-versa) without pointer casting should be safe. A gboolean TRUE is non-zero, so it will be cast to `true` for a bool variable. A gboolean FALSE is zero, so it will be cast to `false`.

kgaillot replied:
I wasn't thinking of g_assert() but g_assert_cmpint(..., ==, false). I understand that mixing is not harmful in this case, but I'm leaning toward enforcing consistency to keep it on developer's minds when they use one or the other.

There was a time when Pacemaker used gboolean almost exclusively, but bool crept in, and I decided to prefer bool except when used with glib functions, mainly because bool is standard, and I prefer not to force public-API-using external code to include glib.h if there's an easy way around it. But I'm second-guessing that decision. I lean toward keeping it, but not mixing the models.

nrwahl2 replied:
Hmm, yes I see I misread. I'm happy to re-push to use g_assert_cmpint(). I'm struggling to think of when it would ever make a practical difference one way or the other, except for consistency. So I have no strong feelings; I defaulted to using what fit semantically.

nrwahl2 replied:
Come to think of it, another reason I went with g_assert_true/false is your comment in #2137 ;)

> I find ! more readable than ==false for boolean comparisons -- otherwise my mind has to backtrack. Also it's consistent with not using ==true.

kgaillot replied:
Ah yes but that doesn't apply because I'll never read the tests again ;)

I'm fine with it either way. The consistency is indeed the only benefit, but consistency can sometimes be a help in a 150K line code base.

nrwahl2 commented on lib/pengine/rules.c:
```diff
@@ -1032,7 +1032,11 @@ pe__eval_attr_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data)
+                    type = "number";
+                } else {
+                    type = "integer";
+                }
```
I think we concluded that backward compatibility should not be an issue, because integers were represented as 32-bit ints and could thus be represented precisely as doubles.

So, if a user has configured type="number" explicitly and they were using numbers that could be represented as 32-bit ints, then representing those numbers as doubles instead (due to explicit "number") should produce the same comparison results.

The same cannot be said for the new long long integer representations after this commit. But any rule expressions that rely on long long above int's capacity would have to be new; there can't be any pre-existing ones.

It could break if the user was previously comparing two unequal values, each greater than or equal to INT_MAX, and expecting them to be equal. But that would be considered a bug in the first place.

kgaillot replied:
Oh right, I forgot. It's good as is.

nrwahl2 commented on doc/Pacemaker_Explained/en-US/Ch-Rules.txt:
```diff
@@ -117,7 +117,12 @@ Expressions are rule conditions based on the values of node attributes.
+ comparison. +number+ performs a floating-point comparison. The default
+ type for +lt+, +gt+, +lte+, and +gte+ operations is +number+ if either
+ value contains a decimal point character, or +integer+ otherwise. The
+ default type for all other operations is +string+.
```
Ah, yep.

nrwahl2 commented:
By the way, we can't rely on GCC and Clang extensions, right? GCC (and I believe Clang as well) has an extension for variadic macro arguments that would eliminate the need to pass NULL as a sentinel value to some of our newer string functions. It's not part of the standard though.

nrwahl2 commented:
I'm having some really frustrating git issues... Every time I go back to rebase and amend, and I resolve a bunch of conflicts, I lose a commit from the log entirely. I don't know whether I'll be able to save this PR or not... I might have to blow it away and start fresh on a new branch.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,123 @@ crm_parse_int(const char *text, const char *default_text)
+ */
+int
+pcmk__scan_double(const char *text, double *result, const char *default_text,
+                  char **end_text)
```
In this iteration, I made pcmk__scan_double internal but non-static. I'm undecided on whether it should be static... It does all the work, so it's a helper function in that regard, but it accepts one argument (end_text) that pcmk__parse_double() does not.

The int parsing functions get by just fine without exposing scan_ll()'s end_text argument outside the file, however. So maybe this ought to be static.

We're accepting a pointer to hold the result and returning success/failure because that's consistent with the usual scan interface. pcmk__parse_double() is designed as a more user-friendly wrapper that returns the parsed value and stores the return code in a pointer. The parsed value is going to be used a lot more often as a function argument than the success/failure code.

nrwahl2 commented on include/crm/common/internal.h:
```diff
@@ -223,6 +225,32 @@ gboolean pcmk__str_in_list(GList *lst, const gchar *s);
+ *       from \c strtod() on other parse errors).
+ */
+static inline double
+pcmk__parse_double(const char *text, const char *default_text, int *rc)
```
Wrapper function for pcmk__scan_double(). The parsed value is usually a lot more interesting to the caller than the rc, so we return the value here. This also passes NULL for end_text.

This is now internal, per your comment in the last push.

nrwahl2 commented on lib/common/strings.c:
```diff
@@ -148,6 +150,123 @@ crm_parse_int(const char *text, const char *default_text)
+    int rc = pcmk_rc_ok;
+    char *local_end_text = NULL;
+
+    CRM_CHECK(result != NULL, return pcmk_rc_no_output);
```
I guessed on some of these return codes. Let me know if any of them don't make sense or if there are better choices. I'm returning all pacemaker return codes now instead of returning errno's.

kgaillot replied:
The standard codes are relatively new, so there is some guesswork involved. I usually scan errno(3) first then do "ack --hh pcmk_rc_" to see existing choices.

In this particular case, EINVAL is the go-to choice for incorrectly supplied arguments. There's nothing wrong with repurposing an existing code if its name sounds reasonably appropriate, but so far pcmk_rc_no_output's use has been related to output in the sense of printing output to the user. Adding new codes is encouraged when a new message would make the user's life easier.

nrwahl2 replied:
I realized after I called it a night that there were probably string descriptions of the pcmk return codes. And indeed there are. Those are helpful, and would have steered me away from pcmk_rc_no_output if I had looked at them :)

nrwahl2 replied:
Are pcmk_rc_{after,before}_range reasonable for differentiating overflow and underflow respectively? Or do you think ERANGE makes more sense for both (as strtod() does)?

kgaillot replied:
Good question, by name either could make sense. However these are the POSIX errnos with their system messages:

* ERANGE: Result too large
* EOVERFLOW: Value too large to be stored in data type

It's interesting that there are two standard errno's for "too large" and none for "too small". There doesn't seem to be a consistently observed difference between the above two, but a decent interpretation is that ERANGE would be for a logically invalid value (e.g. 13 for month), whereas EOVERFLOW is for a logically valid value that can't be represented.

strtod() is an interesting case. It's not defined by POSIX but by C99, which specifies ERANGE as its errno, but perror() on a POSIX system (or equivalently pcmk_rc_str() in pacemaker) would certainly be confusing for underflow.

The pacemaker range codes are similar to the conception of ERANGE as logically invalid, but distinguishing "too small" and "too large", and not necessarily being invalid but just occurring outside the range (e.g. January 5 is not in March through May).

I would say EOVERFLOW is an obvious choice for overflow, but what for underflow? I would say a new pcmk_rc_underflow, or just (ab)use EOVERFLOW and tell callers not to use pcmk_rc_str() if *result is 0.

nrwahl2 replied:
That's enlightening. A quick search for EOVERFLOW vs. ERANGE shows this has indeed been ambiguous for a while. ERANGE seems to have been historically used a lot of places where EOVERFLOW is a better fit, due to the way ERANGE is used in the C standard.

> I would say a new pcmk_rc_underflow, or just (ab)use EOVERFLOW and tell callers not to use pcmk_rc_str() if *result is 0.

It's no trouble to add a pcmk_rc_underflow, so that seems like a better route than abusing EOVERFLOW. (My need for symmetry is dying to add a pcmk_rc_overflow, but then we would be triply redundant ;) ).

nrwahl2 replied:
Done and pushed.

kgaillot commented:
> By the way, we can't rely on GCC and Clang extensions, right? GCC (and I believe Clang as well) has an extension for variadic macro arguments that would eliminate the need to pass NULL as a sentinel value to some of our newer string functions. It's not part of the standard though.

Correct. Exception: glib defines its own wrappers for gcc extensions (like G_GNUC_INTERNAL) that it ensures are usable regardless of compiler, and we could similarly define our own wrappers. But we shouldn't use compiler extensions directly, and some aren't amenable to wrapping.

kgaillot commented:
>  frustrating git issues

Never seen that ;)

git rebase --abort is a good friend of mine (along with git am --abort). The usual culprit is instinctively doing git commit --amend in the middle of resolving conflicts, which merges the conflicting commit into the previous one; you can only do git rebase --continue when resolving conflicts.

nrwahl2 commented:
> git rebase --abort is a good friend of mine (along with git am --abort).

Mine too :)

> The usual culprit is instinctively doing git commit --amend in the middle of resolving conflicts, which merges the conflicting commit into the previous one; you can only do git rebase --continue when resolving conflicts.

I think that's exactly what has been happening. I really appreciate you suggesting this even after I deleted my git-related comment last night (I fixed it then with some tedious cherry-picking from a new branch). I'll watch out for that.

nrwahl2 commented:
> But we shouldn't use compiler extensions directly, and some aren't amenable to wrapping.

Right, this one is `##__VA_ARGS`, which makes it so that the arguments are optional by removing the preceding comma if no arguments are passed. It would allow us to define macros that automatically insert NULL at the end of an argument list, even in the absence of any VA_ARGS. There is a **really** hack-ish way to do it without the extension, but it's such a nightmare to look at that it's not worth introducing for such minimal benefit.

nrwahl2 commented on lib/pengine/rules.c:
```diff
@@ -1043,9 +1047,24 @@ pe__eval_attr_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data)
+            double value_f;
+
+            pcmk__scan_double(h_val, &h_val_f, NULL, NULL);
+            pcmk__scan_double(value, &value_f, NULL, NULL);
```
We may want to check for parse errors here. I didn't do so in the first pass because:
- I haven't given much thought to what we should do if one or both values fail to parse.
- We haven't done any parse error checking in this function prior to this PR, which is already large.

I'm all ears if you have an opinion. I think the simplest approach would be to return false immediately without doing any further processing. Otherwise, there's no **obviously correct** approach for deciding how to handle how comparisons should be handled (e.g., is -3 less than a parse failure?).

nrwahl2 replied:
Checking for parse errors could be considered a backwards compatibility break. On the other hand, you could make the argument that parse failure behavior has been undefined in the past, since the values are expected to be numbers if type="number" is used.

In practice, a parse failure has always yielded the value -1.

nrwahl2 replied:
Note to self: Update/add regression tests if we check for parse errors.

kgaillot replied:
Now you see what I meant about using return codes to force developers to think about error handling : )

I'm thinking a parse error should switch the comparison type to string. At least then we're returning a value that has some meaning (always returning false still assigns an order to parse errors, but that order varies by operation).

nrwahl2 replied:
Good idea, and yep.

nrwahl2 replied:
> always returning false still assigns an order to parse errors

Only in a sense. It would cause the expression to evaluate to false, regardless of the operation. The comparison isn't in a separate function. The comparison and the handling of its result are both in pe__eval_attr_expr.

nrwahl2 commented on lib/pengine/rules.c:
```diff
@@ -1032,7 +1032,11 @@ pe__eval_attr_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data)
+                    type = "number";
+                } else {
+                    type = "integer";
+                }
```
I wonder if we should default to type="string" for lt, lte, gt, and gte if the value doesn't begin with a digit or decimal point, or whether that would be a compatibility concern.

I would think not a compatibility concern, because a user shouldn't expect any particular behavior from strings being compared as numbers.

nrwahl2 replied:
Note to self: Update/add regression tests if we do this.

kgaillot replied:
> I wonder if we should default to type="string" for lt, lte, gt, and gte if the value doesn't begin with a digit or decimal point, or whether that would be a compatibility concern.

Or a minus sign. I'm thinking a detect-text-type function returning 0 = string, 1 = integer, 2 = float. Or it might be easier to just try parsing doubles, and if that fails, try parsing integers, and if that fails it's a string. To be honest, I'd rather just default to string unconditionally, but that would break compatibility.

> I would think not a compatibility concern, because a user shouldn't expect any particular behavior from strings being compared as numbers.

I agree. The current behavior would be unintuitive.

nrwahl2 replied:
> Or it might be easier to just try parsing doubles, and if that fails, try parsing integers, and if that fails it's a string.

I'm just doing "try whichever numeric parse type was selected (if one was selected), and if that fails, then default to string." That will also be documented in Pacemaker Explained. Fixed in next push.

nrwahl2 replied:
I considered the "detect-text-type" function, but the "try X then default to Y on failure" seemed like a more straightforward approach. Since they're fairly lightweight, it's easier to let the parsing functions do our error checking for us instead of bloating the code with checks.

kgaillot commented on lib/common/strings.c:
```diff
@@ -148,6 +150,120 @@ crm_parse_int(const char *text, const char *default_text)
+    char *local_end_text = NULL;
+
+    CRM_ASSERT(result != NULL);
+    *result = -1.0;
```
FYI there is a use case for result==NULL ... the caller could just want to check whether a string would successfully parse or not (checking the validity of command-line arguments or option values is an example). But don't change this, we don't have a use case for it now and it's internal API so we can change it later if needed.

nrwahl2 replied:
I actually considered that, and decided "screw it" for the reasons you mentioned.

kgaillot commented on lib/common/strings.c:
```diff
@@ -41,7 +41,10 @@ crm_itoa_stack(int an_int, char *buffer, size_t len)
+ * \return Standard Pacemaker return code (\c pcmk_rc_ok on success,
+ *         \c EINVAL on failed string conversion due to invalid input,
+ *         or \c ERANGE on arithmetic overflow)
+ * \note Sets \c errno on error
```
I'm leaning against this. The strtoll() errno values could be system-dependent, and using strtoll() is an implementation detail that callers shouldn't rely on, so I think the current description is sufficient. It would be fine to call out ERANGE as a particular value of interest. (Even better would be to return EOVERFLOW or pcmk_rc_underflow as appropriate.)

nrwahl2 replied:
I'm hesitant to use pcmk_rc_underflow in this context, since it seems "underflow" usually refers to a floating-point number with a subnormal absolute value. At a minimum, we'd have to use a more generic string for the return code (e.g, "Value underflowed" instead of "Value too small to be stored in data type").

We could define pcmk_rc_int_underflow and pcmk_rc_float_underflow, but even then there's a lack of consistency. A "too negative" int would be considered an underflow, while a "too negative" double would be considered an overflow. Maybe pcmk_rc_neg_underflow and pcmk_rc_zero_underflow or something...

kgaillot replied:
I was thinking of negative numbers as smaller than positive numbers, so "too small" could make sense as a message. But as you mentioned, it would be inconsistent with doubles.

I'm thinking mainly in terms of pcmk_rc_str() giving a reasonable message to the user. If we think "value too large" is intuitive enough for a "too negative" double, it also makes sense for "too negative" int.

nrwahl2 replied:
> If we think "value too large" is intuitive enough for a "too negative" double

It might not be ideal, but it is the standard EOVERFLOW message. We'll need to add a special case to pcmk_rc_str() for the EOVERFLOW message or define a pcmk_rc_overflow if we don't like it.

FWIW the full message is "Value too large to be stored in data type", which IMO at least **implies** that the absolute value is the important part.

kgaillot replied:
Yeah, I think the situation is less than ideal but acceptable. We can stick with ERANGE or use EOVERFLOW.

nrwahl2 replied:
Switching to EOVERFLOW in next psuh.

kgaillot commented on include/crm/crm.h:
```diff
@@ -51,7 +51,7 @@ extern "C" {
  * >=3.2.0:  DC supports PCMK_LRM_OP_INVALID and PCMK_LRM_OP_NOT_CONNECTED
  */
-#  define CRM_FEATURE_SET		"3.4.1"
+#  define CRM_FEATURE_SET		"3.4.2"
```
This has to be 3.5.0 to have any effect on the cluster (i.e. ensuring a newer node can't be DC until all nodes have been upgraded). The minor-minor is just for external code to be able to test for new features.

nrwahl2 commented:
> Still need to move the default documentation but otherwise looking good

FWIW I **did** fix this, but I must have lost it while recovering commits after botching my history. I'll get it in the next push :)

kgaillot commented:
The giant string PR was merged, hopefully the conflicts will be obvious

nrwahl2 commented on lib/pengine/rules.c:
```diff
@@ -1032,7 +1032,11 @@ pe__eval_attr_expr(xmlNodePtr expr, pe_rule_eval_data_t *rule_data)
     if (value != NULL && h_val != NULL) {
```
Next push will break the comparison logic into two helper functions (one called by the other).

- Gives us an easier way to default to string comparison if numeric parse fails (call the function again with the type argument set to "string").
- Easier to read.

nrwahl2 commented on lib/pengine/rules.c:
```diff
@@ -965,12 +965,163 @@ pe_eval_subexpr(xmlNode *expr, pe_rule_eval_data_t *rule_data, crm_time_t *next_
+    } else if (pcmk__str_eq(op, "ne", pcmk__str_casei)) {
+        return (cmp != 0);
+
+    } else if (l_val == NULL || r_val == NULL) {
```
It seems like we could do without this. At best, it saves us from four string comparisons that follow this block.

compare_attr_expr_vals() treats a NULL value as 0 or an empty string for the sake of comparison. If we eliminate this check, then we just use that result from compare_attr_expr_vals() instead of saying "the whole expression evaluates to false no matter what." It would be a slight change in logic, but the existing behavior was not publicly defined.

kgaillot replied:
> compare_attr_expr_vals() treats a NULL value as 0 or an empty string for the sake of comparison.

Actually NULL always compares as less than non-NULL (the comparison value could be negative).



>  It would be a slight change in logic, but the existing behavior was not publicly defined.

Yes but it's better not to surprise people on upgrades. :)

I think the current approach makes sense from the user's point of view anyway. If a user wants some configuration "when the value of node attribute A is less than some number N", I'm guessing they want that to be considered true only if there is an actual attribute A specified for the node. We'd need a new "value-default" attribute for <expression> to let the user decide what "no attribute exists" should mean, if we wanted to change that. (Which of course we don't want to do unless there is demand for it.)

nrwahl2 replied:
> Actually NULL always compares as less than non-NULL (the comparison value could be negative).

I honestly don't know why I typed that it's treated as zero.. You're right of course and I saw that too.

nrwahl2 commented on lib/pengine/rules.c:
```diff
@@ -965,12 +965,163 @@ pe_eval_subexpr(xmlNode *expr, pe_rule_eval_data_t *rule_data, crm_time_t *next_
+
+    return false;   // Should never reach this point
+}
+
```
May want to break this into two commits (one to refactor into separate functions, and one to do the type logic fix).

nrwahl2 replied:
The push just now is the same content as the one this afternoon, but with the libpe_status commit broken into three commits.

kgaillot commented on lib/pengine/rules.c:
```diff
@@ -1000,15 +1000,25 @@ compare_attr_expr_vals(const char *l_val, const char *r_val, const char *type,
+            int rc1 = errno;
+
+            long long r_val_num = crm_parse_ll(r_val, NULL);
+            int rc2 = errno;
```
I'd say the crm_parse_ll() makes this a feature (for the commit message, i.e. expanding the range of node attribute comparisons).

nrwahl2 replied:
Fixed

kgaillot commented on lib/pengine/rules.c:
```diff
@@ -1044,7 +1071,8 @@ compare_attr_expr_vals(const char *l_val, const char *r_val, const char *type,
  * \param[in]   type    How to interpret the values (allowed values:
- *                      \c "string", \c "number", \c "version", \c NULL)
+ *                      \c "string", \c "number", \c "version",
+ *                      \c "version", \c NULL)
```
I think you meant "integer" for one of these :)

nrwahl2 replied:
Sure did.

nrwahl2 commented:
I just noticed that I updated cts-scheduler instead of cts-scheduler.in for some of the added regression tests. Going to have to rewrite the list additions and re-push.

kgaillot commented:
Awesome :)
